src/service/xhr-impl.js

```diff
@@ -195,9 +176,11 @@ export class Xhr {
    *
    * @param {string} input
    * @param {?FetchInitDef=} opt_init
+   * @param {boolean=} opt_skipProcessing Disables any processing of the
+   *     response, including asserting the response is 2XX.
    * @return {!Promise<!JSONType>}
    */
-  fetchJson(input, opt_init) {
+  fetchJson(input, opt_init, opt_skipProcessing) {
```

Comment:
>  How often do we process errors anyway though?

In 99% of cases, the caller only wants successful JSON. But there's need in A4A for the response headers and JSON (or maybe it's text), and now Forms needs the successful _or_ failure JSON. AMP Access could use it too.

But all of the normal callers don't care about error JSON, so if the JSON decode at the of `xhr.fetchJSON().then(resp => resp.json())` fails because of bad parsing or because of network issues, it's the same "Oh well, request failed" to them. So they keep the same API regardless (just add a `then(res => res.json())`) and their success and failure cases are the same. But for the ones that actually need the Response and it's body, they can take advantage of the more generic API.

>  In principle, this could be extracted from the catch block, and my current WIP code does that, but this complicates the code and undermines type safety, so it's not ideal.

We essentially have 3 cases: Success, Server error, Network Failure. But server error can return Response data (headers, or body), and so it's essentially a success. Having server errors go through the network failure path (`#catch` blocks) is a bad API.

> let's maybe create a separate fetchJsonResponse method and use it in

I'm gonna reject this one, because I've already [ripped out](https://github.com/ampproject/amphtml/pull/7138/files#diff-eccb89b08543ad578ace4060bef469d5L195) a `#fetchJsonResponse` method once. What `#fetchJson` should really do is "setup a JSON XHR message, make the request, and give me the response".

> fetchJson is a different story a bit, since there's no point to parse JSON when there's no content there

This is covered by the above 3 paths I mentioned. To the average caller, failure due to no JSON body (because it's a 404 or whatever) is the same error condition as network failure. So, adding the `then(res => res.json())` will get them to the same spot:

```js
// old API
xhr.fetchJson(url).then(json => {
  // Do stuff
}, () => {
  // Log some user message
});

// New API
xhr.fetchJson(url).then(res => {
  // If network failed, never called. Go straight to #catch block below
  // If response received and theres no body due to 404, then this will go to #catch below
  // Else, go to #then below
  return res.json()
}).then(json => {
  // Do stuff
}, () => {
  // Log some user message
}